jl4-lsp/src/LSP/L4/Viz/VizExpr.hs

```diff
@@ -74,21 +72,13 @@ instance HasCodec BoolValue where
 -- https://github.com/NorfairKing/autodocodec/blob/e939442995debec6d0e014bfcc45449b3a2cb6e6/autodocodec-api-usage/src/Autodocodec/Usage.hs#L688
 -- https://github.com/NorfairKing/autodocodec/blob/e939442995debec6d0e014bfcc45449b3a2cb6e6/autodocodec-api-usage/src/Autodocodec/Usage.hs#L740
 
-instance HasCodec IRDecl where
-  codec = object "IRDecl" $ discriminatedUnionCodec "$type" enc dec
-    where
-      enc (MkFunDecl uid name params body) = ("FunDecl", mapToEncoder (uid, name, params, body) funDeclCodec)
-      dec = HashMap.fromList
-        [ ("FunDecl", ("FunDecl", mapToDecoder mkFunDecl funDeclCodec))
-        ]
-      funDeclCodec =
-        (,,,)
-          <$> requiredField' "id" .= view _1
-          <*> requiredField' "name" .= view _2
-          <*> requiredField' "params" .= view _3
-          <*> requiredField' "body" .= view _4
-
-      mkFunDecl (uid, name, params, body) = MkFunDecl uid name params body
+instance HasCodec FunDecl where
+  codec = object "FunDecl" $
+    typeField "FunDecl" MkFunDecl
+      <*> requiredField' "id"     .= view #id
+      <*> requiredField' "name"   .= view #fnName
+      <*> requiredField' "params" .= view #params
+      <*> requiredField' "body"   .= view #body
```

Comment:
nice! I had already talked about that a few reviews ago, nice that you found a solution for that!

Reply:
yeah was thinking of our convo too!